drivers: sdhc: Support SDIO/SDMMC driver for STM32 using LL APIs

ExaltZephyr opened the pull request:
This PR introduces support for SDIO/SDMMC driver on STM32 using LL APIs, enabling functionality APIs for SDHC host controllers.

This PR includes:

-  replacing usage of SDIO HAL with SDIO/SDMMC LL
- adding support for SDMMC to existing SDHC driver
- general refactoring

**Known issues**
- When running samples/net/wifi/shell sample, Wifi initializes successfully and the shell starts as expected.
issue appear when running "wifi scan" command, the scan command lists available networks correctly but multiple SDIO CMD52 timeouts are raised
Note:
This issue is not related to the changes in this PR since same timeout errors are reproduced on the current main branch.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32_ll.c:
```diff
@@ -0,0 +1,2371 @@
+			data |= ((uint32_t)(*tmp) << 16U);
+			tmp++;
+			data |= ((uint32_t)(*tmp) << 24U);
+			tmp++;
```
Use Zephyr's `UNALIGNED_GET`.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32_ll.h:
```diff
@@ -0,0 +1,247 @@
+	uint32_t CID[4]; /* SD card identification number table */
+
+	uint32_t block_size; /* Block size for SDIO data transfer */
+} sdhc_stm32_ll_handle_t;
```
The useful fields of this should just be inside the `config`/`data` of the `struct device`.

mathieuchopstm reviewed (changes requested):
(Remarks from a rough overlook - I haven't looked at every detail)

This feels like an attempt to provide a *"HAL compatibility layer with 1:1 interface"* which is a perfectly sensible approach on paper. But one would then expect changes in the main driver to be *mostly* mechanical, which is absolutely not the case here; in fact, it's quite the opposite as several layering violations are introduced (c.f. `sdch_stm32_send_csd_and_save_card_configs` for example). That's not even mentioning that `sdhc_stm32_ll.c` is copy-pasted from the ST HAL which *we are trying to get rid of in the first place*.

Another issue is that too many things are done in the same commit, which makes review *extremely* tedious:
* cosmectic changes
* replacing the HAL with *not-the-HAL*
* integrating support for MMC inside the SDIO driver

Those separate changes should be done in separate commits.

---

I see two ways of reworking this into something acceptable:
* keep the "1:1 HAL reimplementation" idea, but done in a proper manner
  * i.e., the main driver changes should become mechanical only
  * bonus points for avoiding *literally 1:1 interface* by adopting some Zephyr-isms (e.g., `int` as error code)
* abandon the HAL entirely and just modify the driver to work with direct LL calls

I am personally inclined towards the second option as I believe it will result in a driver that is much simpler to understand (and it avoids creating a second file too), but will not block an implementation of option 1 (or any other way) if done cleanly.

Additional remarks/comments:

ExaltZephyr commented:
> (Remarks from a rough overlook - I haven't looked at every detail)
> 
> This feels like an attempt to provide a _"HAL compatibility layer with 1:1 interface"_ which is a perfectly sensible approach on paper. But one would then expect changes in the main driver to be _mostly_ mechanical, which is absolutely not the case here; in fact, it's quite the opposite as several layering violations are introduced (c.f. `sdch_stm32_send_csd_and_save_card_configs` for example). That's not even mentioning that `sdhc_stm32_ll.c` is copy-pasted from the ST HAL which _we are trying to get rid of in the first place_.
> 
> Another issue is that too many things are done in the same commit, which makes review _extremely_ tedious:
> 
> * cosmectic changes
> * replacing the HAL with _not-the-HAL_
> * integrating support for MMC inside the SDIO driver
> 
> Those separate changes should be done in separate commits.
> 
> I see two ways of reworking this into something acceptable:
> 
> * keep the "1:1 HAL reimplementation" idea, but done in a proper manner
>   
>   * i.e., the main driver changes should become mechanical only
>   * bonus points for avoiding _literally 1:1 interface_ by adopting some Zephyr-isms (e.g., `int` as error code)
> * abandon the HAL entirely and just modify the driver to work with direct LL calls
> 
> I am personally inclined towards the second option as I believe it will result in a driver that is much simpler to understand (and it avoids creating a second file too), but will not block an implementation of option 1 (or any other way) if done cleanly.
> 
> Additional remarks/comments:

@mathieuchopstm 
Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?

For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?

mathieuchopstm commented:
> @mathieuchopstm Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> 
> For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?

No, what I mean is that the diff should only contain replacement of `HAL_SDIO_xxx` calls with the equivalent function from HAL replacement (which should rather be called e.g., `sdhc_stm32_hal_wrapper.c`). Existing LL API calls in the main driver can remain as they are today - no need to complicate things. (Note that this is only relevant if you want to go the "HAL replacement" route - option 1 in my original message; if you go for the second option, you would use LL API everywhere in a single file)

ExaltZephyr commented:
> > @mathieuchopstm Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> > For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?
> 
> No, what I mean is that the diff should only contain replacement of `HAL_SDIO_xxx` calls with the equivalent function from HAL replacement (which should rather be called e.g., `sdhc_stm32_hal_wrapper.c`). Existing LL API calls in the main driver can remain as they are today - no need to complicate things. (Note that this is only relevant if you want to go the "HAL replacement" route - option 1 in my original message; if you go for the second option, you would use LL API everywhere in a single file)

@mathieuchopstm 
Yes, that is exactly what we are doing in the main driver.
Each API in sdhc_stm32_ll.c is a replacement for a corresponding HAL API.

For example, in sdhc_stm32_rw_extended, we now call
sdhc_stm32_ll_sdio_write_extended instead of HAL_SDIO_WriteExtended.
The logic that was previously inside HAL_SDIO_WriteExtended is now implemented within sdhc_stm32_ll_sdio_write_extended.

the main driver does not contain only one-to-one replacement calls. There are two additional things involved:

- Some refactoring

- Adding SDMMC support

so the question now is how to proceed:

- Keep the refactoring and SDMMC support in this PR, but split them into separate commits, or

- Move the SDMMC support to a separate PR and keep this PR only for HAL API replacement

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -575,33 +874,41 @@ static int sdhc_stm32_get_card_present(const struct device *dev)
+				res = -EIO;
+				sdhc_stm32_log_err_type(config->hsd);
+			}
+		}
```
If no cd gpio is there, we should just assume the card is present. We should not send any commands to the card, if the sd subsystem doesn't expects it.

ExaltZephyr commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -380,6 +404,236 @@ static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+	config->hsd->SdCard.RelCardAdd = cmd->response[0];
+
+	return sdmmc_res;
+}
```
@maass-hamburg 
can you please clarify why this change needed? since we used these APIs before in the old SDIO driver and as we know we can use any API at SDMMC level?

maass-hamburg replied:
instead of having one ore more functions for each command, we could just combine them. this would greatly reduce code size and complexity.

also using the higher functions like `SDMMC_CmdSetRelAdd` is suboptimal, because 
1. they already analyse the response, but that there in zephyr is the responsibility of the sd subsystem and not the sdhc driver.
2. they block internally, instead interrupts with with semaphores could be used instead, so that the cpu could do other things while we wait for the command to finish.


maybe take a look at this driver how to make it more direct https://github.com/zephyrproject-rtos/zephyr/pull/93816

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -617,57 +924,80 @@ static int sdhc_stm32_reset(const struct device *dev)
+		sdhc_stm32_ll_sdio_irq_handler(config->hsd);
+	} else if (IS_ENABLED(CONFIG_SDMMC_STACK)) {
+		sdhc_stm32_ll_irq_handler(config->hsd);
+	}
```
When both are activated, it will try to do the sdio init first, but if there in not a sdio card connected, it will fail and then will try sdmmc after it.

There could be multiple instances of the sdhc one connected to a sdmmc card and one to a sdio card.

Also the sd subsystem with the sdio, sdmmc and mmc stacks is just one possible user of the sdhc. It would also be possible to have something custom connect directly to the sdhc driver without the sd subsystem. Therefore a sdhc driver should not use options of the sd subsystem to assume anything. It should work completly fine also with the sd subsystem being not activated.

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -380,6 +379,225 @@ static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+	 * Zephyr subsys and SDMMC_CmdSetRelAdd
+	 */
+	cmd->response[0] = cmd->response[0] << 16;
+	dev_data->card_relative_addr = cmd->response[0];
```
why save this?

ExaltZephyr commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -380,6 +379,225 @@ static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+			      ((sdmmc_response3 & 0xFFFF0000U) >> 16U));
+		card_block_num = ((DeviceSize + 1U) * 1024U);
+		dev_data->card_log_block_num = card_block_num;
+	}
```
@maass-hamburg 
yes, we need card_log_block_num for operations like read, write, erase..., mainly for validation purposes. We added it because it is used by the HAL.

Although the Zephyr SD subsystem already performs this validation before read/write operations, we may have custom driver that interface directly with the SDHC driver. 
so, if you believe that any custom driver interacting with SDHC must always validate this before sending requests, then we can remove this.
please confirm.

maass-hamburg replied:
> so, if you believe that any custom driver interacting with SDHC must always validate this before sending requests, then we can remove this.

Yes thats the resposibility of the custom driver then. And even if it didn't check it before, the sd card itself will check it and will error out. Then the response has to be checked by the custom driver. the sdhc there also should just return the response and not check it itself.

maass-hamburg reviewed (changes requested):
this driver is still much to high level. Don't do things that are not necessary

ExaltZephyr commented:
> > > @mathieuchopstm Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> > > For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?
> > 
> > 
> > No, what I mean is that the diff should only contain replacement of `HAL_SDIO_xxx` calls with the equivalent function from HAL replacement (which should rather be called e.g., `sdhc_stm32_hal_wrapper.c`). Existing LL API calls in the main driver can remain as they are today - no need to complicate things. (Note that this is only relevant if you want to go the "HAL replacement" route - option 1 in my original message; if you go for the second option, you would use LL API everywhere in a single file)
> 
> @mathieuchopstm Yes, that is exactly what we are doing in the main driver. Each API in sdhc_stm32_ll.c is a replacement for a corresponding HAL API.
> 
> For example, in sdhc_stm32_rw_extended, we now call sdhc_stm32_ll_sdio_write_extended instead of HAL_SDIO_WriteExtended. The logic that was previously inside HAL_SDIO_WriteExtended is now implemented within sdhc_stm32_ll_sdio_write_extended.
> 
> the main driver does not contain only one-to-one replacement calls. There are two additional things involved:
> 
> * Some refactoring
> * Adding SDMMC support
> 
> so the question now is how to proceed:
> 
> * Keep the refactoring and SDMMC support in this PR, but split them into separate commits, or
> * Move the SDMMC support to a separate PR and keep this PR only for HAL API replacement



> Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> 
> For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?



> > (Remarks from a rough overlook - I haven't looked at every detail)
> > This feels like an attempt to provide a _"HAL compatibility layer with 1:1 interface"_ which is a perfectly sensible approach on paper. But one would then expect changes in the main driver to be _mostly_ mechanical, which is absolutely not the case here; in fact, it's quite the opposite as several layering violations are introduced (c.f. `sdch_stm32_send_csd_and_save_card_configs` for example). That's not even mentioning that `sdhc_stm32_ll.c` is copy-pasted from the ST HAL which _we are trying to get rid of in the first place_.
> > Another issue is that too many things are done in the same commit, which makes review _extremely_ tedious:
> > 
> > * cosmectic changes
> > * replacing the HAL with _not-the-HAL_
> > * integrating support for MMC inside the SDIO driver
> > 
> > Those separate changes should be done in separate commits.
> > I see two ways of reworking this into something acceptable:
> > 
> > * keep the "1:1 HAL reimplementation" idea, but done in a proper manner
> >   
> >   * i.e., the main driver changes should become mechanical only
> >   * bonus points for avoiding _literally 1:1 interface_ by adopting some Zephyr-isms (e.g., `int` as error code)
> > * abandon the HAL entirely and just modify the driver to work with direct LL calls
> > 
> > I am personally inclined towards the second option as I believe it will result in a driver that is much simpler to understand (and it avoids creating a second file too), but will not block an implementation of option 1 (or any other way) if done cleanly.
> > Additional remarks/comments:
> 
> @mathieuchopstm Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> 
> For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?

@mathieuchopstm 

> > > @mathieuchopstm Just to make sure we are on the same page: when you mentioned that “main driver changes should become mechanical only”, did you mean that any logic directly using SDMMC APIs should be moved into sdhc_stm32_ll.c?
> > > For example, functions like sdhc_stm32_select_card and sdhc_stm32_send_relative_addr, which currently call SDMMC APIs internally—should be moved to sdhc_stm32_ll.c?
> > 
> > 
> > No, what I mean is that the diff should only contain replacement of `HAL_SDIO_xxx` calls with the equivalent function from HAL replacement (which should rather be called e.g., `sdhc_stm32_hal_wrapper.c`). Existing LL API calls in the main driver can remain as they are today - no need to complicate things. (Note that this is only relevant if you want to go the "HAL replacement" route - option 1 in my original message; if you go for the second option, you would use LL API everywhere in a single file)
> 
> @mathieuchopstm Yes, that is exactly what we are doing in the main driver. Each API in sdhc_stm32_ll.c is a replacement for a corresponding HAL API.
> 
> For example, in sdhc_stm32_rw_extended, we now call sdhc_stm32_ll_sdio_write_extended instead of HAL_SDIO_WriteExtended. The logic that was previously inside HAL_SDIO_WriteExtended is now implemented within sdhc_stm32_ll_sdio_write_extended.
> 
> the main driver does not contain only one-to-one replacement calls. There are two additional things involved:
> 
> * Some refactoring
> * Adding SDMMC support
> 
> so the question now is how to proceed:
> 
> * Keep the refactoring and SDMMC support in this PR, but split them into separate commits, or
> * Move the SDMMC support to a separate PR and keep this PR only for HAL API replacement

@mathieuchopstm 
 can you please advise which approach you’d like us to take so we can proceed accordingly?

mathieuchopstm commented:
> > @mathieuchopstm Yes, that is exactly what we are doing in the main driver. Each API in sdhc_stm32_ll.c is a replacement for a corresponding HAL API.
> > For example, in sdhc_stm32_rw_extended, we now call sdhc_stm32_ll_sdio_write_extended instead of HAL_SDIO_WriteExtended. The logic that was previously inside HAL_SDIO_WriteExtended is now implemented within sdhc_stm32_ll_sdio_write_extended.
> > the main driver does not contain only one-to-one replacement calls. There are two additional things involved:
> > 
> > * Some refactoring
> > * Adding SDMMC support

which should be split out to separate commits

> > so the question now is how to proceed:
> > 
> > * Keep the refactoring and SDMMC support in this PR, but split them into separate commits, or
> > * Move the SDMMC support to a separate PR and keep this PR only for HAL API replacement
> 
> @mathieuchopstm can you please advise which approach you’d like us to take so we can proceed accordingly?

I think a single PR will be easier to maintain for you but I have no opinion on this question.

Refactoring that is not affected by other changes (i.e., preliminary commits - `1~M` below) can be split out and integrated via its own PR if this stalls for too long.

Basically, the PR should be:
* commit(s) 1~M: cleanup
* commit(s) M~N: replace usage of SDIO HAL with SDIO/SDMMC LL
  * either via shim layer - in this case, *the changes should be mostly mechanical*
  * or by using raw LL calls from the driver - the approach I would prefer (and I think @maass-hamburg too)
* commit(s) N~O: add support for SDMMC to existing driver

ExaltZephyr committed:
drivers: sdhc: stm32: fix PM device registration in DEVICE_DT_INST_DEFINE

The PM_DEVICE_DT_INST_GET macro was defined but not passed to
DEVICE_DT_INST_DEFINE, meaning PM callbacks were not properly
registered.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: simplify card presence detection

When no CD GPIO is configured, assume the card is present instead
of sending commands to probe, to avoids sending unexpected commands
that the SD subsystem doesn't expect.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: remove card reset from host reset function

The reset function should reset the host controller only, not the card.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: refactor request error handling and code organization

- Consolidate sdmmc_res error handling into a single check
  at the end of the switch block, removing redundant per-case checks
- Rename goto label 'out' to 'end' in sdhc_stm32_set_io()
- Move DEVICE_API definition after sdhc_stm32_init() for better
  code organization

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: Refactor SDMMC ISR

- Extracted SDMMC interrupt flag clearing into a new helper function.
- Enhanced the event ISR to log errors, call error logging helpers.
- Deferred PM release to ISR only when a DMA-based command is
  successfully launched

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: use data timeout for rw_extended

Replace CONFIG_SD_CMD_TIMEOUT with the actual timeout value
from the sdhc_data structure (data->timeout_ms) for waiting
on DMA completion semaphore.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: Rename SDHC polling config option

Renamed the Kconfig and code references from SDHC_STM32_POLLING_SUPPORT
to SDHC_STM32_DMA_MODE for clarity.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation

Replace STM32 HAL SDIO API calls with custom LL-based wrapper
functions. This removes the dependency on the STM32 HAL SDIO driver.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: add SDMMC support using LL-implementation

Extend the STM32 SDHC driver to support SD memory cards in
addition to SDIO. This adds handling for the following commands:
- SD_SEND_IF_COND (CMD8)
- SD_APP_CMD (CMD55)
- SD_APP_SEND_OP_COND (ACMD41)
- SD_ALL_SEND_CID (CMD2)
- SD_SEND_CSD (CMD9)
- SD_APP_SEND_SCR (ACMD51)
- SD_SET_BLOCK_SIZE (CMD16)
- SD_READ_SINGLE_BLOCK/SD_READ_MULTIPLE_BLOCK (CMD17/CMD18)
- SD_WRITE_SINGLE_BLOCK/SD_WRITE_MULTIPLE_BLOCK (CMD24/CMD25)
- SD_ERASE_BLOCK_START/END/OPERATION (CMD32/CMD33/CMD38)
- SD_SWITCH (CMD6)
- SD_SEND_STATUS (CMD13)

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
dts: bindings: sdhc: stm32: unify SDIO/SDHC under SDMMC binding

Remove st,stm32-sdio devicetree binding and extend
st,stm32-sdmmc to cover the SDHC/SDIO host-controller use case.
This keeps a single STM32 SDMMC hardware compatible in devicetree
and avoids interface-specific compatible selection.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
boards: arduino: update SDMMC compatible

- Update SDMMC nodes on some arduino boards to use
the STM32 SDHC compatible and disable lagacy SDMMC
disk driver.
- Replace `sdhi-on-gpios` with `pwr-gpios` for SDMMC power
control in board devicetree nodes.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
tests: sdmmc: add STM32H747I Disco M7 test configuration

Add configuration and devicetree overlay files
to enable running SDMMC subsystem tests on the STM32H747I
Discovery M7 core.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -633,44 +633,53 @@ static int sdhc_stm32_reset(const struct device *dev)
-	}
-
-	if ((config->hsd->Instance->STA & SDMMC_STA_DCRCFAIL) != 0U) {
+	if ((Instance->STA & SDMMC_STA_DCRCFAIL) != 0U) {
```
While at it, read `STA` only once.

mathieuchopstm reviewed (commented):
(review stopped at 3eb29a6, will continue later - note that comments are made on a commit-by-commit basis so be sure to use that view for things to make sense)

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -633,44 +633,53 @@ static int sdhc_stm32_reset(const struct device *dev)
 
 	HAL_SDIO_IRQHandler(config->hsd);
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	(void)pm_device_runtime_put(dev);
```
There is a comment now but it's a bit ambiguous (*`actually completed`* implies than it should not be done in case of error, even though it is in fact done in case of error)

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -85,68 +69,75 @@ static int sdhi_power_on(const struct device *dev)
-{
-	ARG_UNUSED(hsd);
-	return HAL_OK;
+	dev_data->error_code = SDMMC_ERROR_NONE;
```
`error_code` is no longer reset except if an unknown error occurs?

ExaltZephyr committed:
drivers: sdhc: stm32: refactor error logging

- Replace multiple individual if statements in sdhc_stm32_log_err_type()
with a static const array of error masks and messages.
- Add some error log messages to improve visibility of SDMMC error
conditions.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>
Signed-off-by: Sarah Younis <zephyr@exalt.ps>
Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr commented on drivers/sdhc/sdhc_stm32.c:
Done. Split into a separate commit: [`drivers: sdhc: stm32: refactor error logging`
](https://github.com/zephyrproject-rtos/zephyr/commit/7fdece54c5ff365c79243353e19ea33818a8ad64)

mathieuchopstm commented:
@ExaltZephyr let me know when branch is ready for another round of review :slightly_smiling_face:

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -401,17 +407,12 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
 	}
 
 	(void)pm_device_runtime_get(dev);
-
```
I don't think this newline should be removed

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -435,15 +436,11 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
 		 */
-		cmd->response[0] = cmd->response[0] << 16;
-		break;
+		cmd->response[0] = cmd->response[0] << 16;		break;
```
???

mathieuchopstm commented on drivers/sdhc/sdhc_stm32_ll.c:
```diff
@@ -0,0 +1,1018 @@
+	Init.HardwareFlowControl = data->Init.HardwareFlowControl;
+	Init.ClockDiv = data->Init.ClockDiv;
+
+#if defined(USE_SD_DIRPOL)
```
Where is this from?

ExaltZephyr committed:
drivers: sdhc: stm32: remove unused cmd_index from data struct

Remove unused cmd_index from sdhc_stm32_data structure

Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
drivers: sdhc: stm32: improve code formatting

Fix whitespace, indentation, and alignment issues throughout
the driver

Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -614,35 +621,56 @@ static int sdhc_stm32_reset(const struct device *dev)
+
+	sdhc_stm32_clear_icr_flags(config->hsd->Instance);
+
+	if (data->error_code != 0) {
```
Sorry, that was mistakenly placed in the wrong commit. Fixed now.

etienne-lms replied:
```suggestion
	if (data->error_code != SDMMC_ERROR_NONE) {

etienne-lms commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -373,6 +373,15 @@ static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+ * @param data Optional data transfer descriptor (for read/write commands)
+ *
+ * @return 0 on success, non-zero code on failure.
+ */
```
Is this change useful? The API is already defined in [sdhc.h](https://elixir.bootlin.com/zephyr/v4.3.0/source/include/zephyr/drivers/sdhc.h#L303-L317), as for the other device API handlers.

etienne-lms commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -467,10 +467,18 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
+	 * Defer PM release to ISR only if the DMA-based command was successfully launched.
+	 * Release PM here for all other cases (polling mode, non-read/write opcodes,
+	 * or DMA command launch failure) — the ISR will not fire in those cases.
+	 */
```
Could you also mention this chnage in the commit message for *"drivers: sdhc: stm32: Refactor SDMMC ISR"*?

etienne-lms commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -85,30 +69,30 @@ static int sdhi_power_on(const struct device *dev)
+	uint32_t error_code = dev_data->error_code;
 
-	if (error_code == HAL_SDIO_ERROR_NONE) {
+	if (error_code == SDMMC_ERROR_NONE) {
```
It seems that when this function is called, there is always an error pending.
That said, for consistency, this test makes sense. Or maybe could be replaced by an assertion.

etienne-lms reviewed (commented):
I think some tests and builds could fail on individual commits in this series.
Suggestion for ordering the commits:
- all initial preparatory commits,
- then only introduce new bindings from *"dts: bindings: sdhc: stm32: replace SDIO binding with SDHC"*,
- then *""drivers: sdhc: stm32: add SDMMC support using LL-implementation"* without changing the compatible ID,
- then update the compatible IDs, tree wide (in driver **Kconfig.stm32** + C source + changes in *"boards: arduino: update SDMMC compatible"*)
- then remove the old bindings, from *"dts: bindings: sdhc: stm32: replace SDIO binding with SDHC"*
- last remains *"tests: sdmmc: add STM32H747I Disco M7 test configuration"*

ExaltZephyr commented on dts/bindings/sdhc/st,stm32-sdhc.yaml:
```diff
@@ -0,0 +1,105 @@
+
+description: STM32 SDMMC/SDIO Host Controller
+
+compatible: "st,stm32-sdhc"
```
Thank you for the clarification! We agree that keeping a single compatible (st,stm32-sdmmc) is the cleanest approach here.

But in this case, to make this work, we will have to drop `st,stm32-sdhc` and combine the bindings into [st,stm32-sdmmc.yaml](https://github.com/ExaltZephyr/zephyr/blob/stm32-sdio-ll-support/dts/bindings/mmc/st%2Cstm32-sdmmc.yaml). This means we must merge the new SDHC properties (adding bus: sd to support child nodes) and make legacy-only properties (like disk-name) optional so that both driver architectures can validate correctly.

The choice of which API compiles will then be handled by Kconfig:

```
config SDHC_STM32_SDMMC
    bool "STM32 SDMMC driver support for SD/SDIO"
    default y
    depends on DT_HAS_ST_STM32_SDMMC_ENABLED
    depends on !SDMMC_STM32
```

And for users wanting to migrate to the new SDHC API, they will have to explicitly disable the legacy driver in their configuration:

```
CONFIG_SDMMC_STM32=n
CONFIG_SDHC=y
CONFIG_SDHC_STM32=y 
```
(Note: This Kconfig exclusivity will be just a temporary transition path. Once the legacy disk driver is deleted, we can drop this workaround and change the enablement flow).

Note that modeling the SD/SDIO card as a child node of the SDMMC controller is already used. see for example:

https://github.com/zephyrproject-rtos/zephyr/blob/6cb279c1b4563a0e3cf6d5aa535077d59bfc46d8/boards/arduino/giga_r1/arduino_giga_r1.dtsi#L58-L80


We will do the same for [`stm32h747i_disco_stm32h747xx_m7.overlay`](https://github.com/ExaltZephyr/zephyr/blob/948274ea8f4b4ddfa503f2a64e1ce986984d778f/tests/subsys/sd/sdmmc/boards/stm32h747i_disco_stm32h747xx_m7.overlay), adding it like this:
```
sdhc: &sdmmc1 {
    /* ... */
    sdmmc {
        compatible = "zephyr,sdmmc-disk";
        disk-name = "SD";
        status = "okay";
    };
};
```

ExaltZephyr committed:
doc: migration_guide: document SDHC STM32 renames

Document migration updates for STM32 SD host changes:

- Rename `CONFIG_SDHC_STM32_POLLING_SUPPORT` to
`CONFIG_SDHC_STM32_DMA_MODE` (default y; set to n for polling)
- Rename devicetree compatible `st,stm32-sdio` to `st,stm32-sdmmc`
- Rename `CONFIG_SDHC_STM32_SDIO` to `CONFIG_SDHC_STM32_SDMMC`
- Consolidate `sdhi-on-gpios` into existing `pwr-gpios` for
`st,stm32-sdmmc`.

Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr committed:
modules: fatfs: handle optional STM32 SDMMC disk-name

Handle optional `disk-name` for `st,stm32-sdmmc` entries in
FF_VOLUME_STRS to avoid build failures when the property is not present.

Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

etienne-lms reviewed (commented):
Looks consistent AFAICT.
A few minor case issues (CamelCase/snake_case), worth to wait CI tests result before fixing them.

ExaltZephyr committed:
tests: ram_context_for_isr: exclude nicla vision m7 platform

Exclude arduino_nicla_vision/stm32h747xx/m7 from
application_development.ram_context_for_isr due to IRQ table
registration conflict on this target.

Signed-off-by: Mohammad Odeh <zephyr@exalt.ps>

ExaltZephyr commented on dts/bindings/mmc/st,stm32-sdmmc.yaml:
```diff
@@ -91,3 +90,16 @@ properties:
+    description: |
+        When defined, it indicates that the external SDIO circuit supports 1.8V signaling.
+
+  sdhi-on-gpios:
```
Resolved. I Standardized SDMMC binding/driver/board DTS on pwr-gpios and removed the duplicate property (`sdhi-on-gpios`).

JarmouniA commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -18,3 +18,7 @@ CONFIG_USE_DT_CODE_PARTITION=y
 CONFIG_REGULATOR_FIXED=y
+
+# Use SDHC STM32 driver (disable legacy SDMMC disk driver)
+CONFIG_SDMMC_STM32=n
```
maybe disabling DISK_DRIVERS is a better way, as it will prevent anything subsystem-dependent from building.

ExaltZephyr replied:
Disabling CONFIG_DISK_DRIVERS would disable the whole disk driver menu and skip building all disk drivers (flash, RAM, loopback, FTL, NVMe), not just STM32 SDMMC.
https://github.com/zephyrproject-rtos/zephyr/blob/149c8b1758a8ac6bd20ce230d2b1ae43c483ef30/drivers/disk/Kconfig#L9-L17

ExaltZephyr commented:
Resolved SonarQube issues and switched register access from raw DT address casting to MMIO APIs (`DEVICE_MMIO_*`).

etienne-lms commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -602,36 +610,54 @@ static int sdhc_stm32_reset(const struct device *dev)
 }
 
-void sdhc_stm32_event_isr(const struct device *dev)
+static void sdhc_stm32_clear_icr_flags(SDMMC_TypeDef *Instance)
```
```suggestion
static void sdhc_stm32_clear_icr_flags(SDMMC_TypeDef *instance)

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -81,317 +121,1332 @@ static int sdhi_power_on(const struct device *dev)
+		dev_data->error_code |= SDMMC_ERROR_REQUEST_NOT_APPLICABLE;
+		return -ENOTSUP;
+	}
+
```
remove, this is the responsibility of the overlaying subsystem. the host controller driver should not save any props of the card if possible

ExaltZephyr replied:
@maass-hamburg  right now, in Zephyr, card_erase_blocks only rejects SDIO cards. For normal SD cards, there is no check for erase capability in zephyr subsystem.
so maybe we can keep this check in the driver for now, and remove it later once zephyr adds subsystem support for erase capability? what do you think?

maass-hamburg replied:
According to the sd specification it states that the erase commands are `Mandatory for writable types of cards`. that is the same for cmd24 and 25 (write block). There we are also not checking it. I think the best in that case would just relay on the sd card either error out or timing out.

ExaltZephyr replied:
@maass-hamburg done

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -401,49 +1456,60 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
+	case SD_ERASE_BLOCK_START:
+	case SD_ERASE_BLOCK_END:
+	case SD_ERASE_BLOCK_OPERATION:
+		res = sdhc_stm32_erase(cmd, dev_data, instance);
```
integrate this into the opcodes at the beginning

ExaltZephyr replied:
@maass-hamburg If you agree with my comment on this code review: [cr-comment](https://github.com/zephyrproject-rtos/zephyr/pull/101617#discussion_r3118463765)

then we will not be able to integrate these opcodes at the beginning, right ?

ExaltZephyr replied:
@maass-hamburg done

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -81,317 +121,1332 @@ static int sdhi_power_on(const struct device *dev)
+	if (!dev_data->props.host_caps.vol_180_support) {
 		LOG_ERR("Host does not support 1.8V signaling");
 		return -ENOTSUP;
 	}
```
this check should also be done in the overlaying subsystem

ExaltZephyr replied:
@maass-hamburg  done

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -81,317 +121,1332 @@ static int sdhi_power_on(const struct device *dev)
+	} else {
 		dev_data->sdio_dma_buf = k_aligned_alloc(CONFIG_SDHC_BUFFER_ALIGNMENT,
-							data->blocks * data->block_size);
+							 data->blocks * data->block_size);
```
we should be able to use `data->data` directly. the CONFIG_SDHC_BUFFER_ALIGNMENT should be ensured by the overlaying subsystem, like the sd subsystem already does

ExaltZephyr replied:
@maass-hamburg the sd subsystem in zephyr already supports this, but the sdio subsystem doesn’t. So we added it ourselves for SDIO read/write.

maass-hamburg replied:
unfortunately we got rule 14 in the zephyr coding guidelines

>> Dynamic memory allocation shall not be used

ExaltZephyr replied:
@maass-hamburg   I see
Is this rule mandatory or optional? since I see other zephyr drivers also use dynamic allocation.
the issue that I need this for samples/net/wifi/shell. Without the alignment, Wi-Fi init fails.

What do you suggest?

maass-hamburg replied:
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html:
>> The severity and other references in the table below are for informational purposes only. The listed rules are all required for Zephyr and all new code should comply with the rules listed below.

leave the aligning to the upper levels (subsystem/application)

ExaltZephyr replied:
@maass-hamburg  
Do you mean we should remove this alignment even though it’s not handled in the subsystem yet?
If we do that, WiFi samples will fail in DMA mode. How should we handle this?

ExaltZephyr replied:
@maass-hamburg  
<img width="832" height="129" alt="image" src="https://github.com/user-attachments/assets/34219517-f92c-4ed7-878b-63edb29c40ce" />

Just to note: this alignment isn’t new in this PR—it was already present in the driver before this PR.
@maass-hamburg so what do you think?

maass-hamburg replied:
i would be in favor of removing it

ExaltZephyr replied:
Hi @erwango @mathieuchopstm 

@maass-hamburg suggested to remove the DMA alignment buffer here and use data->data instead, since alignment should be handled by the Zephyr subsystem. I agree with him.
but the problem is that zephyr does not handle alignment for SDIO yet (it handles alignment for SD read/write only not sdio,).
so removing this alignment form our driver will break samples/net/wifi/shell in dma mode, the SDIO read/write buffers are not aligned, which causes Wi-Fi initialization to fail. That’s why this alignment workaround was added earlier in this driver (it’s not new in this PR)

So we have a few options:
1- Remove the alignment buffer as suggested. However, this will break Wi-Fi samples in DMA mode, so we will need to open an issue in the Zephyr subsystem and make this PR dependent on that fix.
2- I can implement the missing alignment handling in the Zephyr SDIO subsystem to address this and push it in this pr (or other pr) if it is ok with you
3- Keep the current workaround in driver for now and remove it later once SDIO alignment is supported in zephyr subsys.

What do you think?

ExaltZephyr committed:
drivers: sdhc: stm32: drop 1.8V support check

Remove 1.8V capability check from
sdhc_stm32_switch_to_1_8v() since this
validation is expected to be handled by
zephyr subsystem.

Signed-off-by: Sara Touqan <zephyr@exalt.ps>

ExaltZephyr commented:
any more comments? @erwango @maass-hamburg @JarmouniA   @etienne-lms      @mathieuchopstm

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -455,9 +449,13 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
+
+		if (sdmmc_res != 0U) {
+			res = -EIO;
+		}
```
Per my previous comment here: https://github.com/zephyrproject-rtos/zephyr/pull/101617#discussion_r2742148092

In commit `Drivers: sdhc: stm32: refactor request error handling and code organization`, I believe this should be:
```c
if (res == 0) /* && sdmmc_res != 0U */ {
    res = -EIO;
}
```
---

BTW: use lowercase in commit title (`Drivers` -> `drivers`)

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -602,36 +610,54 @@ static int sdhc_stm32_reset(const struct device *dev)
+	 * This covers both success and data transfer errors.
+	 */
+	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
+	(void)pm_device_runtime_put_async(dev, K_NO_WAIT);
```
in commit `drivers: sdhc: stm32: Refactor SDMMC ISR`

This stinks - as far as I understand, it ***can*** fail and will leak a device refcount in such cases. The release should rather be done in thread context - presumably in whatever code is waiting on `device_sync_sem`.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -458,10 +458,18 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
+		(void)pm_device_runtime_put(dev);
+	}
 
+	k_mutex_unlock(&dev_data->bus_mutex);
```
in commit `drivers: sdhc: stm32: Refactor SDMMC ISR`

It seems weird to unlock the `bus_mutex` if we just started an operation, which seems to be what the `if()` above checks for.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -84,6 +84,45 @@ static int sdhi_power_on(const struct device *dev)
+
+	{SDMMC_ERROR_BLOCK_LEN_ERR | SDMMC_ERROR_ERASE_SEQ_ERR | SDMMC_ERROR_BAD_ERASE_PARAM |
+		 SDMMC_ERROR_WP_ERASE_SKIP,
+	 "Block or erase sequence error"},
```
in commit `drivers: sdhc: stm32: refactor error logging`

I don't really see the point of merged multiple messages toghether.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -98,41 +137,22 @@ static void sdhc_stm32_log_err_type(SDIO_HandleTypeDef *hsd)
+	if (!known_error) {
+		LOG_ERR("Unknown SDIO Error: 0x%08X", error_code);
 	}
+#endif /* CONFIG_SDHC_LOG_LEVEL > 0 */
```
in commit `drivers: sdhc: stm32: refactor error logging`:

Rather than keeping track of whether *a known error has been seen*, it would be wiser to keep track of *which errors among those reported are known*.

This way, the "unknown error**s**" message can be displayed if there *are* unknown errors, regardless of other known errors.

Also, I would avoid repeating `SDIO Error` to make it clear that all these errors are reported at the same time rather than multiple independent errors. Roughly:
```c
uint32_t errors = HAL_SDIO_GetError(hsd);

if (errors != 0U) {
    LOG_ERR("SDMMC errors occured!");

    for (size_t i = /* ... */) {
        if (errors & sdmmc_errors[i].mask) {
            errors &= ~sdmmc_errors[i].mask;
            LOG_ERR("%s", sdmmc_errors[i].msg);
        }
    }

    if (errors != 0U) {
        LOG_ERR("Unknown error bit(s): 0x%08X", errors);
    }
}
```
(this requires every entry in the `sdmmc_errors` array to contain exactly one bit though)

mathieuchopstm commented on drivers/sdhc/Kconfig.stm32:
```diff
@@ -18,9 +18,9 @@ if SDHC_STM32_SDIO
+	bool "Use polling mode instead of DMA"
 	help
-	  Enables support for SDHC Polling mode on STM32 microcontrollers.
+	  Use polling mode for data transfers instead of DMA mode.
```
In commit `drivers: sdhc: stm32: Rename SDHC polling config option`:

Since we're already breaking backwards compatibility by renaming the option, we might as well align this with basically all other drivers and make the option *enable DMA* rather than disable it.

```
config SDHC_STM32_DMA_MODE
    bool "..."
    default y                         <--- this is the key part
    help
        ....
```

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -28,16 +31,39 @@ BUILD_ASSERT((CONFIG_SDHC_BUFFER_ALIGNMENT % sizeof(uint32_t)) == 0U);
+/* SDMMC wait flags for RX with DBCKEND */
+#define SDMMC_WAIT_RX_DBCKEND_FLAGS                                                                \
+	(SDMMC_FLAG_RXOVERR | SDMMC_FLAG_DCRCFAIL | SDMMC_FLAG_DTIMEOUT | SDMMC_FLAG_DBCKEND |     \
+	 SDMMC_FLAG_DATAEND)
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

perhaps:
```c
#define SDMMC_WAIT_RX_DBCKEND_FLAGS                                                                \
	(SDMMC_WAIT_RX_FLAGS | SDMMC_FLAG_DBCKEND)
```
(which begs the question of whether the define is even needed in the first place)

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -184,43 +208,68 @@ static int sdhc_stm32_sd_init(const struct device *dev)
+	 */
+	if (DT_INST_NUM_CLOCKS(0) > 1) {
+		if (clock_control_get_rate(clk,
+					   (clock_control_subsys_t)(uintptr_t)&config->pclken[1],
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Don't do a double cast

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -184,43 +208,68 @@ static int sdhc_stm32_sd_init(const struct device *dev)
+
+	/* Wait 74 cycles required power up time before starting operations */
+	if (init.ClockDiv != 0) {
+		sdmmc_clk_rate = data->sdmmc_clk / (2U * init.ClockDiv);
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

`init.ClockDiv` -> `config->clk_div`

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -184,43 +208,68 @@ static int sdhc_stm32_sd_init(const struct device *dev)
+	/* Wait 74 cycles required power up time before starting operations */
+	if (init.ClockDiv != 0) {
+		sdmmc_clk_rate = data->sdmmc_clk / (2U * init.ClockDiv);
+		k_msleep(1U + (74U * 1000U / sdmmc_clk_rate));
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

`1000U` -> `USEC_PER_MSEC`?

btw, isn't 74 cycles very short? perhaps `k_busy_wait()` is more appropriate

mathieuchopstm replied:
Basically the division will always return 0 unless `sdmmc_clk_rate <= 74000` (i.e., 74 kHz) which will never happen.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+static void sdhc_stm32_disable_data_interrupts(SDMMC_TypeDef *instance)
 {
-	const struct sdhc_stm32_config *config = dev->config;
+	__SDMMC_DISABLE_IT(instance, SDMMC_IT_DATAEND | SDMMC_IT_DCRCFAIL | SDMMC_IT_DTIMEOUT |
+					     SDMMC_IT_TXUNDERR | SDMMC_IT_RXOVERR |
+					     SDMMC_IT_IDMABTC | SDMMC_IT_TXFIFOHE |
+					     SDMMC_IT_RXFIFOHF);
```
In commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Instead of listing individual IRQ flags, reuse the `#define`s you made earlier? (`SDMMC_xxx_FLAGS`, possibly several OR'ed together)

ExaltZephyr replied:
@mathieuchopstm I cannot since SDMMC_IT_* and SDMMC_FLAG_* definitions don’t share the same values at SDMMC level. For example, SDMMC_IT_DCRCFAIL is not equal to SDMMC_FLAG_DCRCFAIL.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+static uint32_t sdhc_stm32_convert_block_size(uint32_t block_size)
+{
+	/* SDMMC_DATABLOCK_SIZE_<n>B is log2(n) shifted to DBLOCKSIZE bit position */
+	__ASSERT(IS_POWER_OF_TWO(block_size) && block_size >= 1U && block_size <= 16384U,
```
In commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Nit: would be nice to derive bounds from a HW-specific define/property rather than hardcoding them...

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
 
-	uint8_t func = (cmd->arg >> SDIO_CMD_ARG_FUNC_NUM_SHIFT) & 0x7;
-	uint32_t reg_addr = (cmd->arg >> SDIO_CMD_ARG_REG_ADDR_SHIFT) & SDIO_CMD_ARG_REG_ADDR_MASK;
+uint32_t sdhc_stm32_map_resp_type(uint32_t resp_type)
```
In commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Make function `static`.

Nit: I would have sworn a function in the common code already did this...

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+
+		for (size_t i = 0; i < ARRAY_SIZE(r1_errors); i++) {
+			if (resp & r1_errors[i].flag) {
+				dev_data->error_code = r1_errors[i].error;
```
In commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

If we perform an *assignment* here, shouldn't the "log IRQ error" function early abort the moment it sees ONE error bit (since only one can ever be set)?

Nit: seems weird to not use the OCR bits directly as error codes.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+	case SD_RSP_TYPE_R1: {
+		if ((resp & SDMMC_OCR_ERRORBITS) == SDMMC_ALLZERO) {
+			return 0;
+		}
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

I assume `SDMMC_ALLZERO` is zero? If so, then use a plain `0` which is more readable.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+	}
+}
+
+/**
+ * @brief Set response registers into cmd->response[] based on response type.
+ *
+ * @param dev_data  Pointer to the STM32 SDHC driver data structure
+ * @param instance  SDMMC peripheral base address
+ * @param cmd       Zephyr SDHC command structure
+ */
+int sdhc_stm32_set_response(struct sdhc_stm32_data *dev_data, SDMMC_TypeDef *instance,
+			    struct sdhc_command *cmd)
+{
+	uint32_t resp = sdhc_stm32_map_resp_type(cmd->response_type);
+
+	__SDMMC_CLEAR_FLAG(instance, SDMMC_STATIC_CMD_FLAGS);
+	switch (resp) {
+	case SDMMC_RESPONSE_LONG:
+		cmd->response[0] = SDMMC_GetResponse(instance, SDMMC_RESP1);
+		cmd->response[1] = SDMMC_GetResponse(instance, SDMMC_RESP2);
+		cmd->response[2] = SDMMC_GetResponse(instance, SDMMC_RESP3);
+		cmd->response[3] = SDMMC_GetResponse(instance, SDMMC_RESP4);
+		break;
+	case SDMMC_RESPONSE_SHORT:
+		cmd->response[0] = SDMMC_GetResponse(instance, SDMMC_RESP1);
+		return validate_response_err_flags(dev_data, cmd);
+	case SDMMC_RESPONSE_NO:
+	default:
+		break;
+	}
+
+	return 0;
+}
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

If the function cannot fail, make it return `void`.

ExaltZephyr replied:
@mathieuchopstm it can fail if only "validate_response_err_flags" fails

mathieuchopstm replied:
Whoops, I missed this. OK as is then.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+		config->DataLength = blocks * dev_data->block_size;
+		config->DataBlockSize = sdhc_stm32_convert_block_size(dev_data->block_size);
+	} else {
+		config->DataLength = (size_byte > 0U) ? size_byte : 512U;
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Nit: named constant instead of `512U`

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+	if (is_block_mode) {
+		/* Compute number of blocks */
+		uint32_t blocks = (size_byte & ~(dev_data->block_size & 1U)) >>
+				  (find_lsb_set(dev_data->block_size) - 1);
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Nit: the comment is not very useful. either describe how this quite convoluted formula works, or don't say anything at all.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -251,106 +300,619 @@ static int sdhc_stm32_activate(const struct device *dev)
+		} else {
+			data = 0U;
+			for (byte_count = 0U; byte_count < remaining; byte_count++) {
+				data |= ((uint32_t)(*buf) << (byte_count << 3U));
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

`byte_count << 3U` --> `byte_count * BITS_PER_BYTE`

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -368,47 +930,62 @@ static int sdhc_stm32_rw_extended(const struct device *dev, struct sdhc_command
 }
 
-static int sdhc_stm32_switch_to_1_8v(const struct device *dev)
+static int sdhc_stm32_switch_to_1_8v(const struct device *dev, struct sdhc_command *cmd)
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

I assume the `cmd` here is a "switch to 1.8V" command? Should be explained in a comment on top of the function

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -368,47 +930,62 @@ static int sdhc_stm32_rw_extended(const struct device *dev, struct sdhc_command
+	res = sdhc_stm32_send_cmd(dev_data, instance, cmd);
+	if (res == 0) {
+		res = sdhc_stm32_set_response(dev_data, instance, cmd);
+		LOG_DBG("Successfully switched to 1.8V signaling");
```
in commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`:

Can't say that it's successful as you didn't check the return value. The caller should be logging success if anybody(?)

mathieuchopstm commented on drivers/sdhc/Kconfig.stm32:
```diff
@@ -1,18 +1,19 @@
 	default y
 	help
-	  Enable support for SDMMC on STM32 platforms for SDIO devices.
+	  Enables support for SDMMC on STM32 platforms for SD/SDIO devices.
```
In commit `drivers: sdhc: stm32: add SDMMC support using LL-implementation`:

Rework the message:
```
Enables support for SD/SDIO devices using SDMMC peripheral on STM32.
```

mathieuchopstm commented on drivers/sdhc/Kconfig.stm32:
```diff
@@ -1,18 +1,19 @@
-	bool "STM32 SDMMC driver support for SDIO"
-	depends on DT_HAS_ST_STM32_SDIO_ENABLED
+config SDHC_STM32_SDMMC
+	bool "STM32 SDMMC driver support for SD/SDIO"
```
In commit `drivers: sdhc: stm32: add SDMMC support using LL-implementation`:

`bool "STM32 SDMMC host controller driver` (not sure about the `host controller` part - maybe omit it)

mathieuchopstm commented on drivers/sdhc/Kconfig.stm32:
```diff
@@ -1,18 +1,19 @@
+config SDHC_STM32_SDMMC
+	bool "STM32 SDMMC driver support for SD/SDIO"
+	depends on DT_HAS_ST_STM32_SDMMC_ENABLED
+	depends on !SDMMC_STM32
```
In commit `drivers: sdhc: stm32: add SDMMC support using LL-implementation`:

A comment explaining why this is needed would be appreciated.

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -559,6 +564,278 @@ int sdhc_stm32_set_response(struct sdhc_stm32_data *dev_data, SDMMC_TypeDef *ins
+		ret = sdhc_stm32_rw_blocks_poll(dev_data, cmd, instance, data->data, data->blocks,
+						data->timeout_ms, is_write);
+	}
+
+	if (!IS_ENABLED(CONFIG_SDHC_STM32_POLLING_MODE) && ret == 0) {
+		if (k_sem_take(&dev_data->device_sync_sem, K_MSEC(data->timeout_ms)) != 0) {
+			LOG_ERR("Failed to acquire Semaphore");
+			return -ETIMEDOUT;
+		}
```
In commit `drivers: sdhc: stm32: add SDMMC support using LL-implementation`:

you should probably kill the transfer in this case

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
In commit `drivers: sdhc: stm32: add SDMMC support using LL-implementation`:

overall, there seems to be many functions lacking `static`

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
Nit: couldn't commit `drivers: sdhc: stm32: drop 1.8V support check` be done earlier?

mathieuchopstm commented on dts/bindings/mmc/st,stm32-sdmmc.yaml:
Outside PR scope(?): this should probably be removed from the binding.

mathieuchopstm commented on boards/arduino/giga_r1/arduino_giga_r1.dtsi:
For atomicity, **the DTS changes** from commit `boards: arduino: update SDMMC compatible` should probably be squashed with the commit updating the binding.

mathieuchopstm reviewed (commented):
Seems fine overall, with a few cosmetic changes/cleanups still needed here and there. On the functional side, a few points are worrying me - see review comments.

mathieuchopstm commented:
@maass-hamburg don't hesitate to close your old comments / `Hide` > `Outdated` to make the PR thread more readable :slightly_smiling_face:

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -460,8 +460,8 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
 
 	pm_policy_state_lock_put(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
 	(void)pm_device_runtime_put(dev);
-	k_mutex_unlock(&dev_data->bus_mutex);
 
+	k_mutex_unlock(&dev_data->bus_mutex);
```
in commit `drivers: sdhc: stm32: Refactor SDMMC ISR`

seems to be a stray change that can be dropped

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -96,43 +138,24 @@ static int sdhi_power_on(const struct device *dev)
-	if ((error_code & SDMMC_ERROR_WRITE_PROT_VIOLATION) != 0U) {
-		LOG_ERR("Attempt to program a write protected block");
-	}
+#if CONFIG_SDHC_LOG_LEVEL > 0
```
in commit `drivers: sdhc: stm32: refactor error logging`

Nit:
```
#if CONFIG_SDHC_LOG_LEVEL >= LOG_LEVEL_ERR
```

mathieuchopstm reviewed (commented):
A few more things:

maass-hamburg commented on modules/fatfs/zephyr_fatfs_config.h:
```diff
@@ -107,14 +107,16 @@
 	DT_FOREACH_STATUS_OKAY(zephyr_sdmmc_disk, _FF_DISK_NAME)                                   \
 	DT_FOREACH_STATUS_OKAY(zephyr_mmc_disk, _FF_DISK_NAME)                                     \
-	DT_FOREACH_STATUS_OKAY(st_stm32_sdmmc, _FF_DISK_NAME)                                      \
+	DT_FOREACH_STATUS_OKAY(st_stm32_sdmmc, _FF_DISK_NAME_IF_PRESENT)                           \
```
propably better to use IF_ENABLED(CONFIG_SDMMC_STM32, ..)

maass-hamburg commented on doc/releases/migration-guide-4.5.rst:
```diff
@@ -103,6 +103,32 @@ GPIO
+  .. code-block:: kconfig
+
+     CONFIG_SDMMC_STM32=n
+     CONFIG_SDHC_STM32_SDMMC=y
```
CONFIG_SDHC_STM32_SDMMC doesn't need to be enabled explicitly, as it is enabeled by default if CONFIG_SDMMC_STM32=n

maass-hamburg commented on boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7_defconfig:
```diff
@@ -18,3 +18,7 @@ CONFIG_USE_DT_CODE_PARTITION=y
+
+# Use SDHC STM32 driver (disable legacy SDMMC disk driver)
+CONFIG_SDMMC_STM32=n
+CONFIG_SDHC_STM32_SDMMC=y
```
here too

maass-hamburg commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -401,49 +1463,54 @@ static int sdhc_stm32_request(const struct device *dev, struct sdhc_command *cmd
+		res = sdhc_stm32_send_cmd(dev_data, instance, cmd);
+		if (res == 0) {
+			res = sdhc_stm32_set_response(dev_data, instance, cmd);
 		}
```
can be combined

mathieuchopstm commented on drivers/sdhc/sdhc_stm32.c:
```diff
@@ -240,79 +286,667 @@ static int sdhc_stm32_activate(const struct device *dev)
+
+	__ASSERT(IS_POWER_OF_TWO(block_size) && block_size >= 1U && block_size <= max_blk_size,
+		"Invalid block size: %u", block_size);
+	return LOG2(block_size) << SDMMC_DCTRL_DBLOCKSIZE_Pos;
```
In commit `drivers: sdhc: stm32: replace HAL_SDIO with LL-based implementation`

The comment should be moved right above the `return ...;` line

mathieuchopstm reviewed (dismissed):
Non-blocking remark, otherwise LGTM. I'll let @maass-hamburg take a look again and give the second approval 🙂

Thank you very much for the hard work!

maass-hamburg commented on drivers/sdhc/Kconfig.stm32:
```diff
@@ -1,26 +1,29 @@
 	select SDHC_SUPPORTS_NATIVE_MODE
 	select PINCTRL
-	imply CACHE_MANAGEMENT if !SDHC_STM32_POLLING_SUPPORT && DCACHE
+	imply CACHE_MANAGEMENT if SDHC_STM32_DMA_MODE && DCACHE
```
this could now be moved to 
``config SDHC_STM32_DMA_MODE``,

also it should probably be selected, because if there is a data cache, we need these chache operations

maass-hamburg commented on tests/subsys/sd/sdmmc/boards/stm32h747i_disco_stm32h747xx_m7.overlay:
```diff
@@ -0,0 +1,29 @@
+	status = "okay";
+	min-bus-freq = <DT_FREQ_K(400)>;
+	max-bus-freq = <DT_FREQ_M(208)>;
+	/delete-property/ disk-name;
```
deleting it is no longer needed, as we now do IF_ENABLED(CONFIG_SDMMC_STM32, (DT_FOREACH_STATUS_OKAY(st_stm32_sdmmc, _FF_DISK_NAME)))

schweigstill commented:
I just created #108916.
